### PR TITLE
Archival integration

### PR DESCRIPTION
**Pull request by aaclan-ebi**

### CODE REVIEW

File: src/main/java/org/humancellatlas/ingest/analysis/web/AnalysisController.java [hunk 1]
```diff
@@ -51,6 +53,25 @@
...
+    @RequestMapping(path = "/submissionEnvelopes/{sub_id}/analyses",
+            method = RequestMethod.PUT,
+            produces = MediaTypes.HAL_JSON_VALUE)
+    ResponseEntity<Resource<?>> linkAnalysesToEnvelope(@PathVariable("sub_id") SubmissionEnvelope submissionEnvelope,
+                                                     @RequestBody MetadataReference analysisReference,
+                                                     final PersistentEntityResourceAssembler assembler) {
+        SubmissionEnvelope entity = getAnalysisService().resolveAnalysisReferencesForSubmission(submissionEnvelope, analysisReference);
+        PersistentEntityResource resource = assembler.toFullResource(entity);
+        return ResponseEntity.accepted().body(resource);
+    }
```
Comment (tburdett) [inline src/main/java/org/humancellatlas/ingest/analysis/web/AnalysisController.java]: @aaclan-ebi is your archival code making lots of use of this style of endpoint? I think it's a little problematic, cos we're effectively making a single endpoint (`/submissionEnvelopes/{sub_id}/analyses`) have two different styles of behaviour depending on whether you send a `POST` or `PUT` request.

There's also a standard way to do this - see https://docs.spring.io/spring-data/rest/docs/current/reference/html/#repository-resources.association-resource.

If you really need this to make your code work, we can keep for now, but I much prefer the first style: `path = "/submissionEnvelopes/{sub_id}/analyses/{id}", method = RequestMethod.PUT`

Comment (aaclan-ebi) [reply src/main/java/org/humancellatlas/ingest/analysis/web/AnalysisController.java]: I see, okies, will change this. I also added a logic not to be able to link to a new submission env if it's still "open". Using the standard way may ignore that logic so I'll go with the first style.

File: src/main/java/org/humancellatlas/ingest/analysis/web/AnalysisController.java [hunk 2]
```diff
@@ -72,7 +93,7 @@
     ResponseEntity<Resource<?>> addFileReference(@PathVariable("analysis_id") Analysis analysis,
                                                  @RequestBody File file,
                                                  final PersistentEntityResourceAssembler assembler) {
-        SubmissionEnvelope submissionEnvelope = analysis.getSubmissionEnvelope();
+        SubmissionEnvelope submissionEnvelope = analysis.getLatestSubmissionEnvelope();
         file.addToSubmissionEnvelope(submissionEnvelope);
         File entity = getFileRepository().save(file);
         Analysis result = getAnalysisService().getAnalysisRepository().save(analysis.addFile(entity));
```
Comment (tburdett) [inline src/main/java/org/humancellatlas/ingest/analysis/web/AnalysisController.java]: I can imagine a situation where this doesn't work very well (e.g. if you try to add a file to an analysis whose latest submission is complete).

I *think* the submission envelope must be open to add a file to an analysis - so therefore maybe we should just change this to `POST /submissionEnvelopes/{sub_id}/analyses/{analysis_id}/files`

File: src/main/java/org/humancellatlas/ingest/assay/web/AssayController.java [hunk 3]
```diff
@@ -49,6 +51,26 @@
...
+            method = RequestMethod.PUT,
+            produces = MediaTypes.HAL_JSON_VALUE)
+    ResponseEntity<Resource<?>> linkAssaysToEnvelope(@PathVariable("sub_id") SubmissionEnvelope submissionEnvelope,
+                                                       @RequestBody MetadataReference assayReference,
+                                                       final PersistentEntityResourceAssembler assembler) {
+        SubmissionEnvelope entity = getAssayService().resolveAssayReferencesForSubmission(submissionEnvelope, assayReference);
+        PersistentEntityResource resource = assembler.toFullResource(entity);
+        return ResponseEntity.accepted().body(resource);
+    }
+
```
Comment (tburdett) [inline src/main/java/org/humancellatlas/ingest/assay/web/AssayController.java]: See my comments for analysis, above - same drill here

File: src/main/java/org/humancellatlas/ingest/analysis/web/AnalysisController.java (hunk 2 above)
Comment (aaclan-ebi) [reply src/main/java/org/humancellatlas/ingest/analysis/web/AnalysisController.java]: I think the linking of files to analysis shouldn't be affected. This just gets the last in the list, which should be still currently open when linking files to analysis

File: src/main/java/org/humancellatlas/ingest/core/MetadataDocument.java [hunk 4]
```diff
@@ -42,17 +47,30 @@ protected MetadataDocument(EntityType type,
     }
 
     public MetadataDocument addToSubmissionEnvelope(SubmissionEnvelope submissionEnvelope) {
-        this.submissionEnvelope = submissionEnvelope;
-
+        if(!this.isLinkedToOpenSubmissionEnvelope()){
+            this.validationState = ValidationState.DRAFT;
```
Comment (tburdett) [inline src/main/java/org/humancellatlas/ingest/core/MetadataDocument.java]: Don't set state manually here.  Instead use `enactStateTransition(ValidationState.DRAFT)` which will throw an error is the state transition isn't an allowed one (adds some extra safety)

File: src/main/java/org/humancellatlas/ingest/core/MetadataDocument.java [hunk 5]
```diff
@@ -42,17 +47,30 @@ protected MetadataDocument(EntityType type,
     }
 
     public MetadataDocument addToSubmissionEnvelope(SubmissionEnvelope submissionEnvelope) {
-        this.submissionEnvelope = submissionEnvelope;
-
+        if(!this.isLinkedToOpenSubmissionEnvelope()){
```
Comment (tburdett) [inline src/main/java/org/humancellatlas/ingest/core/MetadataDocument.java]: I really like this check, very nice

File: src/main/java/org/humancellatlas/ingest/core/MetadataDocument.java [hunk 6]
```diff
@@ -101,4 +121,12 @@ public MetadataDocument enactStateTransition(ValidationState targetState) {
...
         return this;
     }
+
+    @JsonIgnore
+    public SubmissionEnvelope getLatestSubmissionEnvelope(){
+        if(this.submissionEnvelopes.size() > 0){
+            return this.submissionEnvelopes.get(this.submissionEnvelopes.size() - 1);
+        }
+        return null;
+    }
```
Comment (tburdett) [inline src/main/java/org/humancellatlas/ingest/core/MetadataDocument.java]: I think we should change this - I *think* you are always trying to get the open submission here, right?

If so, why not change this method to `getOpenSubmissionEnvelope()` and loop over the submission envelopes, looking for one that is still open (i.e. not yet submitted).  Bonus points: add an `isOpen()` check to `SubmissionEnvelope`.

**Comment by aaclan-ebi** [comment]
All comments applied now and merged to master.

**Pull request merged**